Kubernetes/legos/k8s_get_memory_utilization_of_services/k8s_get_memory_utilization_of_services.py

```diff
@@ -118,39 +82,78 @@ def k8s_get_memory_utilization_of_services(handle, namespace: str = "", threshol
-        else:
-            service_pods = " ".join(services)
-            kubectl_command = f"kubectl get pods -n {namespace} {service_pods} -o json"
+        # Get memory utilization of all pods using top pods
+        top_pods_command = f"kubectl top pods -n {namespace} --no-headers"
+        response = handle.run_native_cmd(top_pods_command)
+        top_pods_output = response.stdout.strip() 
+        if not top_pods_output:
+            raise ValueError(f"Top PODS data is empty for given namespace {namespace}")
+        top_pods_output = top_pods_output.split('\n')
+        pod_mem_util_dict = {x.split()[0]: x.split()[-1] for x in top_pods_output}
+
+        pods_to_check = {}
+        if services:
+            # If services specified, lets iterate over it and get pods corresponding to them.
+            # If service pod not found in the top pod list, which means the memory
+            # utilization is not significant, so dont need to check
+            for svc in services:
+                kubectl_cmd = f"kubectl get service {svc} -n {namespace} -o=jsonpath={{.spec.selector}}"
+                response = handle.run_native_cmd(kubectl_cmd)
+                svc_labels = json.loads(response.stdout.strip())
+                if not svc_labels:
+                    continue
+                _labels = ", ".join([f"{key}={value}" for key, value in svc_labels.items()])
+                svc_pod_cmd = f"kubectl get pods -n {namespace} -l \"{_labels}\" -o=jsonpath={{.items[*].metadata.name}}"
+                response = handle.run_native_cmd(svc_pod_cmd)
+                svc_pods = response.stdout.strip()
+                if not svc_pods:
+                    # No pods attached to the given service
+                    continue
+                for svc_pod in svc_pods.split():
+                    if svc_pod in pod_mem_util_dict.keys():
+                        pods_to_check[svc_pod] = pod_mem_util_dict[svc_pod]
+
+        if not pods_to_check:
+            pods_to_check = pod_mem_util_dict
+
+        for pod, mem_usage in pods_to_check.items():
+            # Lets get memory requests for all containers not just the first container
+            kubectl_command = f"kubectl get pod {pod} -n {namespace} -o=jsonpath='{{.spec.containers[*].resources.requests.memory}}'"
             response = handle.run_native_cmd(kubectl_command)
-            json_data = json.loads(response.stdout.strip())
-        
-        pod_data = parse_pod_data(json_data)
-        for data in pod_data: 
-            mem_usage = convert_memory_to_bytes(data.get('memory_request')) 
-            mem_limit = convert_memory_to_bytes(data.get('memory_limit'))
-            if not mem_limit:
-                # Memory limit is not set, so lets continue with the execution
+            mem_request = response.stdout.strip()
+            _mem_request = 0
+            for _mr in mem_request.split():
+                # Lets calculate the total memory request for all containers in the POD
```

Comment:
oh why are we adding the total memory to calculate the total memory of the pod??? How does kubectl top pod work? Does it aggregate the memory usage of all the containers?
I saw this option ```kubectl top pods --containers``` which gives containers usage.

Reply:
I think this is fine 
```
Each container in the Pod gets its own individual limit and request, but because Pods are always scheduled as a group, you need to add the limits and requests for each container together to get an aggregate value for the Pod.
```